numpy/_core/src/multiarray/stringdtype/casts.cpp

```diff
@@ -658,6 +852,135 @@ pyobj_to_string(PyObject *obj, char *out, npy_string_allocator *allocator)
+        loop_descrs[1] = given_descrs[1];
+    }
+
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
+
+    return NPY_UNSAFE_CASTING;
+}
+
+// Example template parameters:
+// TNpyType: npy_int8
+// TNpyLongType: npy_longlong
+// typenum: NPY_BYTE
+template <typename TNpyType, typename TNpyLongType, NPY_TYPES typenum>
+static int string_to_int(
+    PyArrayMethod_Context * context,
+    char *const data[],
+    npy_intp const dimensions[],
+    npy_intp const strides[],
+    NpyAuxData *NPY_UNUSED(auxdata)
+) {
+    PyArray_StringDTypeObject *descr =
+            ((PyArray_StringDTypeObject *)context->descriptors[0]);
+    npy_string_allocator *allocator =
+            NpyString_acquire_allocator(descr);
+    int has_null = descr->na_object != NULL;
+    const npy_static_string *default_string = &descr->default_string;
+
+    npy_intp N = dimensions[0];
+    char *in = data[0];
+    TNpyType *out = (TNpyType *)data[1];
+
+    npy_intp in_stride = strides[0];
+    npy_intp out_stride = strides[1] / sizeof(TNpyType);
+
+    while (N--) {
+        TNpyLongType value;
+        if (stringbuf_to_int<TNpyLongType>(in, &value, has_null, default_string, allocator) != 0) {
+            npy_gil_error(PyExc_RuntimeError, "Encountered problem converting string dtype to integer dtype.");
+            goto fail;
+        }
+        *out = (TNpyType)value;
+
+        // Cast back to TNpyLongType to check for out-of-bounds errors
+        if (static_cast<TNpyLongType>(*out) != value) {
+            // out of bounds, raise error following NEP 50 behavior
+            const char *errmsg = NULL;
+            if constexpr (std::is_same_v<TNpyLongType, npy_ulonglong>) {
+                errmsg = "Integer %llu is out of bounds for %s";
+            } else if constexpr (std::is_same_v<TNpyLongType, npy_longlong>) {
+                errmsg = "Integer %lli is out of bounds for %s";
+            } else {
+                errmsg = "Unrecognized integer type %i is out of bounds for %s";
+            }
+            npy_gil_error(PyExc_OverflowError, errmsg, value, typenum_to_cstr(typenum));
+            goto fail;
+        }
+        in += in_stride;
+        out += out_stride;
+    }
+
+    NpyString_release_allocator(allocator);
+    return 0;
+
+  fail:
+    NpyString_release_allocator(allocator);
+    return -1;
+}
+
+template<typename TNpyType, typename TNpyLongType, NPY_TYPES typenum>
+static PyType_Slot s2int_slots[] = {
+    {NPY_METH_resolve_descriptors, (void *)&string_to_int_resolve_descriptors<typenum>},
+    {NPY_METH_strided_loop, (void *)&string_to_int<TNpyType, TNpyLongType, typenum>},
+    {0, NULL}
+};
+
+static const char *make_s2type_name(NPY_TYPES typenum) {
+    const char *prefix = "cast_StringDType_to_";
+    size_t plen = strlen(prefix);
+
+    const char *type_name = typenum_to_cstr(typenum);
+    size_t nlen = strlen(type_name);
+
+    char *buf = (char *)PyMem_RawCalloc(sizeof(char), plen + nlen + 1);
```

Comment:
ugh, sorry, this must have been annoying to deal with. Thankfully I think you're right that there only needs to be one spot where this gets freed, in `dtype.c` when we're done setting up the casts.